src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/InsertOperationPage.java

```diff
@@ -15,59 +19,118 @@
+  enum InsertionStatus {
+    OK,
+    ERROR,
+    SKIPPED
+  }
+
+  @JsonPropertyOrder({"_id", "status", "errorsIdx"})
+  @JsonInclude(JsonInclude.Include.NON_NULL)
+  record InsertionResult(DocumentId _id, InsertionStatus status, Integer errorsIdx) {}
 
   /** No-arg constructor, usually used for aggregation. */
-  public InsertOperationPage() {
-    this(new ArrayList<>(), new HashMap<>());
+  public InsertOperationPage(
+      List<InsertOperation.WritableDocAndPosition> allAttemptedInsertions,
+      boolean returnDocumentResponses) {
+    this(allAttemptedInsertions, returnDocumentResponses, new ArrayList<>(), new ArrayList<>());
   }
 
   /** {@inheritDoc} */
   @Override
   public CommandResult get() {
-    // if we have errors, transform
-    if (null != failedIds && !failedIds().isEmpty()) {
-
-      List<CommandResult.Error> errors = new ArrayList<>(failedIds.size());
-      failedIds.forEach((documentId, throwable) -> errors.add(getError(documentId, throwable)));
+    // Ensure insertions and errors are in the input order (wrt unordered insertions),
+    // regardless of output format
+    Collections.sort(successfulInsertions);
+    if (!failedInsertions().isEmpty()) {
+      Collections.sort(
+          failedInsertions, Comparator.comparing(tuple -> tuple.getItem1().position()));
+    }
 
+    if (!returnDocumentResponses()) { // legacy output, limited to ids, error messages
+      List<CommandResult.Error> errors;
+      if (failedInsertions().isEmpty()) {
+        errors = null;
+      } else {
+        errors =
+            failedInsertions.stream()
+                .map(tuple -> getError(tuple.getItem1().document().id(), tuple.getItem2()))
+                .toList();
+      }
+      // Old style, simple ids:
+      List<DocumentId> insertedIds =
+          successfulInsertions.stream().map(docAndPos -> docAndPos.document().id()).toList();
       return new CommandResult(null, Map.of(CommandStatus.INSERTED_IDS, insertedIds), errors);
     }
 
-    // id no errors, just inserted ids
-    return new CommandResult(Map.of(CommandStatus.INSERTED_IDS, insertedIds));
+    // New style output: detailed responses.
+    InsertionResult[] results = new InsertionResult[allAttemptedInsertions().size()];
+    List<CommandResult.Error> errors = new ArrayList<>();
+
+    // Results array filled in order: first successful insertions
+    for (InsertOperation.WritableDocAndPosition docAndPos : successfulInsertions) {
+      results[docAndPos.position()] =
+          new InsertionResult(docAndPos.document().id(), InsertionStatus.OK, null);
+    }
+    // Second: failed insertions
+    for (Tuple2<InsertOperation.WritableDocAndPosition, Throwable> failed : failedInsertions) {
+      InsertOperation.WritableDocAndPosition docAndPos = failed.getItem1();
+      Throwable throwable = failed.getItem2();
+      CommandResult.Error error = getError(throwable);
+      int errorIdx = errors.indexOf(error);
+      if (errorIdx < 0) { // not yet added, add
```

Comment:
So this part is to reuse the same error, reuse the same errorIdx.
Nitpick, maybe good to add one line comment, if I do not read the spec previously, may have trouble understanding this `errorIdx = errors.size();`

Reply:
Ok will add a comment.